smart-hadoop-support/smart-hadoop/src/main/java/org/smartdata/hdfs/action/DistCpAction.java

```diff
@@ -0,0 +1,145 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.smartdata.hdfs.action;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.hadoop.mapreduce.Cluster;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.tools.DistCp;
+import org.apache.hadoop.tools.DistCpOptions;
+import org.apache.hadoop.tools.OptionsParser;
+import org.smartdata.action.Utils;
+import org.smartdata.action.annotation.ActionSignature;
+
+@ActionSignature(
+    actionId = "distcp",
+    displayName = "DistCp",
+    usage = DistCpAction.FILE_PATH + " $path "
+        + DistCpAction.TARGET_ARG + " $target "
+        + " [additional options from "
+        + "https://hadoop.apache.org/docs/stable/hadoop-distcp/DistCp.html#Command_Line_Options]"
+)
+public class DistCpAction extends HdfsAction {
+
+  public static final String TARGET_ARG = "-target";
+  public static final String SOURCE_PATH_LIST_FILE = "-f";
+
+  private static final String SOURCE_PATHS_DELIMITER = ",";
+
+  private String sourcePaths;
+
+  private String targetPath;
+
+  private Map<String, String> distCpArgs;
+
+  @Override
+  public void init(Map<String, String> args) {
+    super.init(args);
+
+    sourcePaths = args.get(FILE_PATH);
+    targetPath = args.get(TARGET_ARG);
+
+    distCpArgs = new HashMap<>(args);
+    distCpArgs.remove(FILE_PATH);
+    distCpArgs.remove(TARGET_ARG);
+  }
+
+  @Override
+  protected void execute() throws Exception {
+    // we need to perform validation and option parsing here
+    // because SSM doesn't correctly handle exceptions thrown in the init method
+    DistCpOptions distCpOptions = buildDistCpOptions();
+
+    DistCp distCp = new DistCp(getContext().getConf(), distCpOptions);
+    appendLog(
+        String.format("DistCp Action started at %s for options %s",
+            Utils.getFormatedCurrentTime(), distCpOptions));
+
+    try (JobCloseableWrapper jobWrapper = new JobCloseableWrapper(distCp.execute())) {
+      distCp.waitForJobCompletion(jobWrapper.job);
+      appendLog(jobWrapper.job.getCounters().toString());
+    }
+  }
+
+  DistCpOptions buildDistCpOptions() {
+    validateActionArguments();
+
+    List<String> rawArgs = distCpArgs.entrySet()
+        .stream()
+        .flatMap(entry -> mapOptionToStr(entry.getKey(), entry.getValue()))
+        .collect(Collectors.toList());
+
+    if (!distCpArgs.containsKey(SOURCE_PATH_LIST_FILE)) {
+      rawArgs.addAll(parseSourcePaths(sourcePaths));
+    }
+    rawArgs.add(targetPath);
+
+    return OptionsParser.parse(rawArgs.toArray(new String[0]));
+  }
+
+  private List<String> parseSourcePaths(String sourcePaths) {
+    return Arrays.asList(sourcePaths.split(SOURCE_PATHS_DELIMITER));
```

Comment:
Add check for emptiness

Reply:
Nice catch! Added checks and tests both for empty source and target args